README.rst
Velha 3D
========
 |Velha|

Jogo para aprendizado de programação.

 |docs| |python| |license| |github|


:Author:  Carlo E. T. Oliveira
:Version: 20.11
:Affiliation: Universidade Federal do Rio de Janeiro
:License: GNU General Public License v3 or later (GPLv3+)
:Homepage: `Jogo da Velha`_
:Changelog: `CHANGELOG <CHANGELOG.rst>`_

Jogo da Velha em 3D
------

Um jogo de tabuleiro como um projeto exemplo para aprender a programar na linguagem Python.

Este projeto facilita a aprendizagem da linguagem Python.

O jogo é dirigido principalmente ao ensino de programação de computadores para jovens e crianças do ensino médio e fundamental.

-------

Laboratório de Automação de Sistemas Educacionais
-------------------------------------------------

**Copyright © Carlo Olivera**

LABASE_ - NCE_ - UFRJ_

|LABASE|

.. _LABASE: http://labase.activufrj.nce.ufrj.br
.. _NCE: http://nce.ufrj.br
.. _UFRJ: http://www.ufrj.br

.. _Jogo da Velha: https://activufrj.nce.ufrj.br/wiki/carlo/Jogo_da_Velha

.. |github| image:: https://img.shields.io/github/v/release/cetoli/velha3d
   :target: https://github.com/cetoli/velha3d/releases/


.. |LABASE| image:: https://cetoli.gitlab.io/spyms/image/labase-logo-8.png
   :target: http://labase.activufrj.nce.ufrj.br
   :alt: LABASE

.. |Velha| image:: https://github.com/cetoli/velha3d/blob/master/source/_static/velha2d.png
   :alt: VELHA 3D

.. |python| image:: https://img.shields.io/github/languages/top/cetoli/velha3d
   :target: https://www.python.org/downloads/release/python-383/

.. |docs| image:: https://img.shields.io/readthedocs/velha3d
   :target: https://velha3d.readthedocs.io/en/latest/index.html

.. |license| image:: https://img.shields.io/github/license/cetoli/velha3d
   :target: https://github.com/cetoli/velha3d/blob/master/LICENSE



CHANGELOG.rst
VELHA 3D - Changelog
====================

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog`_.


`Unreleased`_
-------------
- to be defined

`20.11.b1`_
------------

Added
+++++
- Novas classes.

`20.11.b0`_
------------

Added
+++++
- Adiciona classe Casa.

`20.11.a1`_
-----------

Added
+++++
- Programando com classes.
- Class Velha.



-------

Laboratório de Automação de Sistemas Educacionais
-------------------------------------------------

**Copyright © Carlo Olivera**

LABASE_ - NCE_ - UFRJ_

|LABASE|

.. _LABASE: http://labase.activufrj.nce.ufrj.br
.. _NCE: http://nce.ufrj.br
.. _UFRJ: http://www.ufrj.br
.. _Keep a Changelog: https://keepachangelog.com/en/1.0.0/
.. _20.11.a1: https://github.com/cetoli/velha3d/tree/release20_11_a1
.. _20.11.b0: https://github.com/cetoli/velha3d/tree/release_20_11_b0
.. _20.11.b1: https://github.com/cetoli/velha3d/tree/release_20_11_b1

.. |LABASE| image:: https://cetoli.gitlab.io/spyms/image/labase-logo-8.png
   :target: http://labase.activufrj.nce.ufrj.br
   :alt: LABASE




src/test/testvelha.py
__author__ = 'carlo'

import sys
sys.path.append("/home/carlotolla/dev/velha3d/src/test/")
import unittest
import unittest.mock as mock
import glow


class TestIntegerArithmeticTestCase(unittest.TestCase):
    def tearDown(self):
        self.s.reset_mock()
        self.b.reset_mock()
        self.c.reset_mock()

    @mock.patch('glow.sphere')
    @mock.patch('glow.box')
    @mock.patch('glow.canvas')
    def setUp(self, canvas, box, sphere):
        self.s = sphere
        self.b = box
        self.c = canvas
        from velha import main, Casa
        self.casas = Casa.CASAS

        self.t = main()

    def testClicou(self):
        self.b.reset_mock()
        self.assertIn((0, 0, 0), self.casas, "Not in CASAS:" % self.casas)
        self.t[0].clicou()
        self.t[4].clicou()
        self.t[9].clicou()
        self.t[5].clicou()
        self.t[18].clicou()
        self.assertEqual(len(self.s.mock_calls), 6, "Not 2 spheres: %s" % self.s.mock_calls)
        self.assertEqual(len(self.b.mock_calls), 5, "Not 5 boxes: %s" % self.b.mock_calls)

    def testInit(self):  # test method names begin 'test*'
        #self.assertEqual(len(self.c.mock_calls), 2, "Not 2 canvas: %s" % self.c.mock_calls)
        #self.assertEqual(len(self.b.mock_calls), 27, "Not 27 boxes: %s" % self.b.mock_calls)
        self.assertIn((0, 0, 0), self.casas, "Not in CASAS:" % self.casas)

if __name__ == "__main__":
    unittest.main()
    print('didit', __name__)


src/octo.py
# -*- coding: utf-8 -*-
from glow import *
TAM = (-3, -1, 0, 1, 3)
SP = 2
TAM = (SP * i for i in TAM)
SZ = 4
print("oi")
# equação geral da reta y = a*x + b
# equação da nossa reta (b=0) y = a*x
# melhore o código de TIRAS para que teste mais possibilidades de ganhar
TIRAS = [[(x*SP, y*SP, z*SP) for x in TAM] for y in TAM for z in TAM]+[
        [(x*SP, y*SP, z*SP) for z in TAM] for x in TAM for y in TAM]+[
    [(x*SP, y*SP, z*SP) for y in TAM] for x in TAM for z in TAM]+[
        [(-9,9,z),(0,0,z),(9,-9,z)] for z in [-9,0,9]]+[
        [(9,9,z),(0,0,z),(-9,-9,z)] for z in [-9,0,9]]


CASAS = [(a, 0, 0) for a in [-SP, SP]] + [(0, a, 0) for a in [-SP, SP]] +  [(0, 0, a) for a in [-SP, SP]]
SP *=2
CASAS += [(a, 0, 0) for a in [-SP, SP]] + [(0, a, 0) for a in [-SP, SP]] +  [(0, 0, a) for a in [-SP, SP]]



pecas = [box, sphere] * 14
cores = [color.red, color.blue] * 14


class Casa:
    CASAS = {}  # esta coleção serve para achar o objeto casa a partir de sua posicão

    def __init__(self, x, y, z):
        self.pos = (x*SP, y*SP, z*SP)
        self.e_casa = sphere(pos=self.pos, size=(SZ, SZ, SZ), opacity=0.2)
        Casa.CASAS[self.pos] = self  # adiciona esta casa na coleção de casas
        self.peca = None

    def recebe(self, algo3d):
        self.peca = algo3d
        vencedores = self.testa_ganhou()
        if vencedores:
            print("ganhou")
            self.pinta_vencedores(vencedores)
        return algo3d

    def tipo_peca(self):
        #print("tipo_peca", self.peca.tipo if self.peca is not None else 0)
        return self.peca.tipo if self.peca else 0

    def clicou(self):
        coluna, linha, camada = self.pos  # aposição da peça vai ser a posição da casa
        peca = Peca(pecas.pop(), coluna, linha, camada, cores.pop())  # cria uma peça aqui
        #Casa.CASAS.pop(self.pos)  # remove esta da lista de casas para não ser clicada
        self.recebe(peca)  # avisa a casa que ela esta é a peça que está nela
        print(self.pos)

    def pinta_vencedores(self, vencedores):
        SP = SZ+1
        for vencedor in vencedores:
            for posicao in vencedor:
                print("box", posicao)
                box(pos=posicao, color=color.yellow,  size=(SP, SP, SP), opacity=0.3)
                # modifique para ser um box amarelo transparente do tamanho da casa

    def testa_ganhou(self):
        def casas_ganhadoras(tira):
            tipo_tira = [Casa.CASAS[casa].tipo_peca() for casa in tira if isinstance(casa, tuple)]
            return tipo_tira == [1, 1, 1] or tipo_tira == [2, 2, 2]
        tiras = [tira for tira in TIRAS if casas_ganhadoras(tira)]  # crie aqui um teste para saber se alguem venceu
        #print("testa_ganhou", tiras,  casas_ganhadoras(tiras))
        return tiras


class Peca:
    def __init__(self, tipo_peca, x, y, z, cor):
        self.e_peca = tipo_peca(pos=(x, y, z), color=cor,  size=(SZ, SZ, SZ), opacity=0.6)
        self.tipo = 1 if tipo_peca == box else 2


def main():

    def clicou(event):
        cc = cena.mouse.pick().pos  # pega a posição do objeto clicado pelo mouse
        casa_clicada = (cc.x, cc.y, cc.z)  # cria uma tripla ordenada no espaço
        if casa_clicada in Casa.CASAS.keys():  # procura a tripla na coleção de casas
            casa_clicada = Casa.CASAS[casa_clicada]
            if casa_clicada in TABULEIRO:
                TABULEIRO.remove(casa_clicada)
                casa_clicada.clicou()  # chama o clicou da casa escolhida
    _gs = glow('main')
    cena = canvas()
    cena.width = 400
    cena.height = 400
    cena.bind("mousedown", clicou)

    TABULEIRO = [Casa(coluna, linha, camada)
                 for coluna , linha, camada in CASAS]
    return TABULEIRO

if __name__ == "__main__":
    main()


velha/main.py
# -*- coding : utf-8 -*-
# SPDX-License-Identifier: GPL-3.0-or-later
""" Jogo para ensino de programação Python.

.. codeauthor:: Carlo Oliveira <carlo@ufrj.br>

Changelog
---------
.. versionadded::    20.11
        version.

"""
casa_0, casa_1, casa_2 = None, None, None
"""Primeira linha do tabuleiro"""
casa_3, casa_4, casa_5 = None, None, None
"""Segunda linha do tabuleiro"""

casa_6, casa_7, casa_8 = None, None, None
"""Terceira linha do tabuleiro"""

jogador = "x"
"""O jogador com a peça x começa"""

for turno in range(0):

    tabuleiro = """
Escolha uma casa do tabuleiro
 {} | {} | {}
-----------
 {} | {} | {}
-----------
 {} | {} | {}
""".format(casa_0 or 0, casa_1 or 1, casa_2 or 2,
           casa_3 or 3, casa_4 or 4, casa_5 or 5,
           casa_6 or 6, casa_7 or 7, casa_8 or 8)   
    jogada = str(input(tabuleiro))
    if jogada == "0": casa_0 = jogador
    elif jogada == "1": casa_1 = jogador
    elif jogada == "2": casa_2 = jogador
    elif jogada == "3": casa_3 = jogador
    elif jogada == "4": casa_4 = jogador
    elif jogada == "5": casa_5 = jogador
    elif jogada == "6": casa_60 = jogador
    elif jogada == "7": casa_7 = jogador
    elif jogada == "8": casa_8 = jogador
    jogador = "o" if jogador == "x" else "x"


velha/classe.py
# -*- coding : utf-8 -*-
# SPDX-License-Identifier: GPL-3.0-or-later
""" Jogo para ensino de programação Python.

.. codeauthor:: Carlo Oliveira <carlo@ufrj.br>

Changelog
---------
.. versionadded::    20.11
        Usando uma classe.

"""


class Velha():
    """Representa o jogo da velha.
    
    Referências:
    
    casa
     Lista contendo as casas do jogo.
     
    jogador
     Marcador do jogador com a vez corrente.
    """
    ESCOLHA = "Escolha uma casa do tabuleiro"
    """Frase para pedir uma jogada."""
    FINAL = "Situação final do tabuleiro"
    """Frase indicando o fim o do jogo."""
    TABULEIRO = """
        {}
        {} | {} | {}
        -----------
        {} | {} | {}
        -----------
        {} | {} | {}
        """   
    """Display mostrando a situação atual do tabuleiro."""

    def __init__(self):
        self.casa = [None]*9
        """A operação de multiplicar com a lista contendo um **None**
        gera uma lista contendo nove posições contendo None.
        """

        self.jogador = "x"
        """O jogador com a peça x começa"""
        
    def joga(self):
        """Executa o joga da velha.
        
        Referências:
        
        jogada
        Índice da casa a ser preenchida.
        
        display_das_casas
        Apresentação do conteúdo de cada casa: o índice se vazia ou marcador se cheia.
        """

        while True:
            display_das_casas = [a_casa or posicao for posicao, a_casa in enumerate(self.casa)]

            if all(self.casa):
                print(Velha.TABULEIRO.format(Velha.FINAL, *display_das_casas))
                """Mostra a situação final do tabuleiro."""
                break
            jogada = str(input(Velha.TABULEIRO.format(Velha.ESCOLHA, *display_das_casas)))
            """Mostra a situação atual do tabuleiro e pede a próxima jogada."""
            casa_escolhida = self.casa[int(jogada)]
            """Obtem a peça que está colocada na casa escolhida"""
            if casa_escolhida:
                """Se a casa escolhida já está ocupada volta a fazer a pergunta."""
                continue
            self.casa[int(jogada)] = self.jogador
            """ Coloca a peça na posição pedida pela jogada."""
            self.jogador = "o" if self.jogador == "x" else "x"
            """Troca a vez para o outro jogador."""


if __name__ == '__main__':
    Velha().joga()

velha/função.py
# -*- coding : utf-8 -*-
# SPDX-License-Identifier: GPL-3.0-or-later
""" Jogo para ensino de programação Python.

.. codeauthor:: Carlo Oliveira <carlo@ufrj.br>

Changelog
---------
.. versionadded::    20.11
        Usando uma função.

.. versionchanged::    20.11
        Melhorando a documentação da função.

"""
ESCOLHA = "Escolha uma casa do tabuleiro"
"""Frase para pedir uma jogada."""
FINAL = "Situação final do tabuleiro"
"""Frase indicando o fim o do jogo."""
TABULEIRO = """
    {}
    {} | {} | {}
    -----------
    {} | {} | {}
    -----------
    {} | {} | {}
    """   
"""Display mostrando a situação atual do tabuleiro."""


def velha():
    """Executa o joga da velha.
    
    Referências:
    
    casa
     Lista contendo as casas do jogo.
     
    jogador
     Marcador do jogador com a vez corrente.
    
    jogada
     Índice da casa a ser preenchida.
    
    display_das_casas
     Apresentação do conteúdo de cada casa: o índice se vazia ou marcador se cheia.
    """
    casa = [None]*9
    """A operação de multiplicar com a lista contendo um **None**
    gera uma lista contendo nove posições contendo None.
    """

    jogador = "x"
    """O jogador com a peça x começa"""

    while True:
        display_das_casas = [a_casa or posicao for posicao, a_casa in enumerate(casa)]

        if all(casa):
            print(TABULEIRO.format(FINAL, *display_das_casas))
            """Mostra a situação final do tabuleiro."""
            break
        jogada = str(input(TABULEIRO.format(ESCOLHA, *display_das_casas)))
        """Mostra a situação atual do tabuleiro e pede a próxima jogada."""
        casa_escolhida = casa[int(jogada)]
        """Obtem a peça que está colocada na casa escolhida"""
        if casa_escolhida:
            """Se a casa escolhida já está ocupada volta a fazer a pergunta."""
            continue
        casa[int(jogada)] = jogador
        """ Coloca a peça na posição pedida pela jogada."""
        jogador = "o" if jogador == "x" else "x"
        """Troca a vez para o outro jogador."""


if __name__ == '__main__':
    velha()

velha/classesvec.py
# -*- coding : utf-8 -*-
# SPDX-License-Identifier: GPL-3.0-or-later
""" Jogo para ensino de programação Python.

.. codeauthor:: Carlo Oliveira <carlo@ufrj.br>

Changelog
---------
.. versionadded::    20.11.b0
        Usando uma classe para Casa e uma para Velha.

"""
class Casa():
    """Representa uma casa do jogo da velha.
    
    Referências:
    
    casa
     Célula contendo o marcador do jogador.
    """     

    def __init__(self):
        self.casa = None
        """Uma casa vazia contendo None.
        """
        
    def joga(self, jogador):
        """Preecha a casa com o maracdor de um jogador.
        
        :param jogador: Marcador do jogador para preencher a casa.
        
        """
        self.casa = jogador
        
    def mostra(self):
        """Retorna o marcador do jogador.
        
        :return: Marcador do jogador usado para preencher a casa.
        
        """
        return self.casa


class Velha():
    """Representa o jogo da velha.
    
    Referências:
    
    casa
     Lista contendo as casas do jogo.
     
    jogador
     Marcador do jogador com a vez corrente.
    """
    ESCOLHA = "Escolha uma casa do tabuleiro"
    """Frase para pedir uma jogada."""
    FINAL = "Situação final do tabuleiro"
    """Frase indicando o fim o do jogo."""
    TABULEIRO = """
        {}
        {} | {} | {}
        -----------
        {} | {} | {}
        -----------
        {} | {} | {}
        """   
    """Display mostrando a situação atual do tabuleiro."""

    def __init__(self):
        self.casa = [Casa() for _ in range(9)]
        """Usamos uma compressão de lista para criar nove instâncias da classe **Casa**.
        """

        self.jogador = "x"
        """O jogador com a peça x começa"""
        
    def joga(self):
        """Executa o joga da velha.
        
        Referências:
        
        jogada
         Índice da casa a ser preenchida.
        
        display_das_casas
         Apresentação do conteúdo de cada casa: o índice se vazia ou marcador se cheia.
        """

        while True:
            display_das_casas = [a_casa.mostra() or posicao for posicao, a_casa in enumerate(self.casa)]

            if all([marcador.mostra() for marcador in self.casa]):
                print(Velha.TABULEIRO.format(Velha.FINAL, *display_das_casas))
                """Mostra a situação final do tabuleiro."""
                break
            jogada = str(input(Velha.TABULEIRO.format(Velha.ESCOLHA, *display_das_casas)))
            """Mostra a situação atual do tabuleiro e pede a próxima jogada."""
            casa_escolhida = self.casa[int(jogada)].mostra()
            """Obtem a peça que está colocada na casa escolhida"""
            if casa_escolhida:
                """Se a casa escolhida já está ocupada volta a fazer a pergunta."""
                continue
            self.casa[int(jogada)].joga(self.jogador)
            """ Coloca a peça na posição pedida pela jogada."""
            self.jogador = "o" if self.jogador == "x" else "x"
            """Troca a vez para o outro jogador."""


if __name__ == '__main__':
    Velha().joga()

src/velha.py
# -*- coding: utf-8 -*-
from glow import *
from __random import choice
from browser import doc
#from urllib import unquote
def unquote(txt):
    return txt.replace("&quot;", "\"")
TAM = (-1, 0, 1)
SP = 9
SZ = 4
print("oi")
# equação geral da reta y = a*x + b
# equação da nossa reta (b=0) y = a*x
# melhore o código de TIRAS para que teste mais possibilidades de ganhar
TIRAS = [[(x*SP, y*SP, z*SP) for x in TAM] for y in TAM for z in TAM]+[
        [(x*SP, y*SP, z*SP) for z in TAM] for x in TAM for y in TAM]+[
    [(x*SP, y*SP, z*SP) for y in TAM] for x in TAM for z in TAM]+[
        [(-9,9,z),(0,0,z),(9,-9,z)] for z in [-9,0,9]]+[
        [(9,9,z),(0,0,z),(-9,-9,z)] for z in [-9,0,9]]+[
        [(x,-9,9),(x,0,0),(x,9,-9)] for x in [-9,0,9]]+[
        [(x,9,9),(x,0,0),(x,-9,-9)] for x in [-9,0,9]]+[
        [(-9,y,9),(0,y,0),(9,y,-9)] for y in [-9,0,9]]+[
        [(9,y,9),(0,y,0),(-9,y,-9)] for y in [-9,0,9]]

TIRASD = {
    casa: [tiras for tiras in TIRAS if casa in tiras]
    for casa in [(x*SP, y*SP, z*SP) for x in TAM for y in TAM for z in TAM]}


pecas = [box, sphere] * 14
cores = [color.red, color.blue] * 14


class Casa:
    CASAS = {}  # esta coleção serve para achar o objeto casa a partir de sua posicão

    def __init__(self, x, y, z):
        self.pos = (x*SP, y*SP, z*SP)
        self.e_casa = sphere(pos=self.pos, size=(SZ, SZ, SZ), opacity=0.2)
        Casa.CASAS[self.pos] = self  # adiciona esta casa na coleção de casas
        self.peca = None

    def recebe(self, algo3d):
        self.peca = algo3d
        vencedores = self.testa_ganhou()
        if vencedores:
            print("ganhou")
            #self.pinta_vencedores(vencedores)
            TABULEIRO.ganhou(vencedores)
            return True
        return False

    def tipo_peca(self):
        #print("tipo_peca", self.peca.tipo if self.peca is not None else 0)
        return self.peca.tipo if self.peca else 0

    def clicou(self):
        coluna, linha, camada = self.pos  # aposição da peça vai ser a posição da casa
        TABULEIRO.registra_jogada_na_historia(self.pos)
        #peca = Peca(pecas.pop(), coluna, linha, camada, cores.pop())  # cria uma peça aqui
        peca = TABULEIRO.joga(coluna, linha, camada)  # cria uma peça aqui
        #Casa.CASAS.pop(self.pos)  # remove esta da lista de casas para não ser clicada
        return self.recebe(peca)  # avisa a casa que ela esta é a peça que está nela
        #print(self.pos)

    def testa_ganhou(self):
        def casas_ganhadoras(tira):
            tipo_tira = [Casa.CASAS[casa].tipo_peca() for casa in tira if isinstance(casa, tuple)]
            return tipo_tira == [1, 1, 1] or tipo_tira == [2, 2, 2]
        tiras = [tira for tira in TIRASD[self.pos] if casas_ganhadoras(tira)]  # crie aqui um teste para saber se alguem venceu
        #print("testa_ganhou", tiras,  casas_ganhadoras(tiras))
        return tiras


class Peca:
    def __init__(self, tipo_peca, x, y, z, cor):
        self.e_peca = tipo_peca(pos=(x, y, z), color=cor,  size=(SZ, SZ, SZ), opacity=0.6)
        self.tipo = 1 if tipo_peca == box else 2

    def esconde(self):
        self.e_peca.visible = False

    def move_e_mostra(self, x, y, z):
        self.e_peca.visible = True
        self.e_peca.pos = (x, y, z)
        return self


class Tabuleiro:
    # um dicionário que coleta movimentos ganhadores.
    # O valor de cada chave é uma lista de tuplas onde cada tupla é:
    # (número de vitórias com essa jogada, índice da jogada em Tabuleiro.CASAS)
    # a chave é uma string com a lista de jogadas até agora separadas por espaço em branco
    # exemplo: {'_':[(1, 1)],'_ 1':[],'_ 1 8':[(1, 18)],'_ 1 8 18 17':[(1, 26)]}
    HISTORY = {'_':[]}
    CASAS = []
    def __init__(self, cena):
        self.jogada_robo = (0, 0, 0)
        self.peca_robo = 2
        self.cena = cena
        self._casas = [Casa(coluna, linha, camada)
                 for coluna in TAM for linha in TAM for camada in TAM]
        Tabuleiro.CASAS = [(coluna*SP, linha*SP, camada*SP)
                 for coluna in TAM for linha in TAM for camada in TAM]
        self._pecas = [Peca(peca, 0, 0, 0, cores.pop()) for peca in pecas]
        SM = SZ+1
        self._marcadores = [box(pos=(0, 0, 0), color=color.yellow,
            size=(SM, SM, SM), opacity=0.3) for x in range(9)]
        #print([casa.tipo_peca() for casa in Casa.CASAS.values()])
        self.inicia()

    def registra_jogada_na_historia(self, jogada):
        #cria uma chave que é um texto com os indices das casas jogadas separadas por brancos
        self.jogadas += " %d" % Tabuleiro.CASAS.index(jogada)
        # recupera as vitórias na história referentes a estas jogadas, inicia com [] se não acha
        self.vitorias_nesta_jogada = Tabuleiro.HISTORY.setdefault(self.jogadas, [])  # {"0":0, "1":0, "2":0})

    def remove(self, casa):
        self.casas.remove(casa)

    def inicia(self, ev=0):
        print(Tabuleiro.HISTORY)
        self.jogadas = "_"
        self.vitorias_nesta_jogada = Tabuleiro.HISTORY["_"]
        self._clica = self.clicou
        self.casas = self._casas[:]
        self.pecas = self._pecas[:]
        self.marcadores = self._marcadores[:]
        for casa in self.casas:
            casa.peca = None
        for peca in self.pecas:
            peca.esconde()
        for marcador in self.marcadores:
            marcador.visible = False
        if TABULEIRO and self.peca_robo == 2:
            self.jogada()
        # alterna a vez de iniciar entre o humano e o robô
        self.peca_robo = 1 if self.peca_robo == 2 else 2
        #print([casa.tipo_peca() for casa in Casa.CASAS.values()])
        #print([casa.tipo_peca() for casa in self.casas])

    def joga(self, x, y, z):
        return self.pecas.pop().move_e_mostra(x, y, z)

    def verifica_alguem_ganha(self, casa_anterior, pc=1):
        def casas_ganhadoras(tira):
            tipo_tira = [Casa.CASAS[casa].tipo_peca() for casa in tira if isinstance(casa, tuple)]
            return tipo_tira == [pc, pc, 0] or tipo_tira == [0, pc, pc] or tipo_tira == [pc, 0, pc]
        tiras = [tira for tira in TIRASD[casa_anterior] if casas_ganhadoras(tira)]  # crie aqui um teste para saber se alguem venceu
        #print("testa_ganhou", tiras,  casas_ganhadoras(tiras))
        return tiras

    def jogada(self, casa_anterior=(0, 0, 0)):
        def casa_do_indice(ind):
            return Casa.CASAS[Tabuleiro.CASAS[ind]]
        casa_da_jogada = choice(self.casas)  # choice(TABULEIRO)  # TABULEIRO[0]
        alguem_ganha = self.verifica_alguem_ganha(self.jogada_robo,  1 if self.peca_robo == 2 else 2)
        if not alguem_ganha:
            alguem_ganha = self.verifica_alguem_ganha(casa_anterior, self.peca_robo)
        if alguem_ganha:
            casas = alguem_ganha[0]
            casa_vazia = [Casa.CASAS[casa] for casa in casas if Casa.CASAS[casa].tipo_peca() == 0]
            print('humano_ganha:', alguem_ganha, casas, "jogue aqui:", casa_vazia)

            casa_da_jogada = casa_vazia[0]
        # AI: consulta a hitória de vitórias para repetir um lance vitorioso
        if self.vitorias_nesta_jogada and choice(range(10)) >-1:
            # seleciona entre os palpites vitoriosos um que caia numa casa vazia
            palpite = [casa_do_indice(tentativa[1])
                for tentativa in self.vitorias_nesta_jogada
                if casa_do_indice(tentativa[1]) in self.casas]
            # se exite um palpite possível seleciona o primeiro da lista ordenada
            if palpite:
                casa_da_jogada = palpite[0]
        self.remove(casa_da_jogada)
        casa_da_jogada.clicou()
        self.jogada_robo = casa_da_jogada.pos
        return self.jogada_robo

    def clica(self, event):
        self._clica(event)

    def ganhou(self, vencedores):
        for vencedor in vencedores:
            for posicao in vencedor:
                print("box", posicao)
                marca = self.marcadores.pop()
                marca.pos = posicao
                marca.visible = True
        self._clica = self.inicia
        jogadas = self.jogadas.split()
        l = len(jogadas)-1
        #AI : cria uma lista de chaves que representam as jogadas que levaram à vitória
        ganhadoras = [(" ".join(jogadas[:l-i]), int(jogadas[l-i]))
            for i in range(0, len(jogadas), 2)
            if not '_' in jogadas[l-i]]
        print( ganhadoras, jogadas, "-%s-"%self.jogadas, jogadas[:-0])
        #AI : atualiza a história adicionando uma vitória a cada tupla de jogada da lista
        for chave, jogada in ganhadoras:
            # recupera o registro de vitórias representado por esta chave
            registros = Tabuleiro.HISTORY[chave]
            # acha a tupla cuja jogada é a jogada usada nesta vitória
            registro = [par for par in registros if jogada == par[1]]
            # remove esta tupla se existia, para que fique duplicada
            if registro:
                registros.remove(registro[0])
            # cria ou recupera a tupla vitoriosa
            ganhos, casa = registro[0] if registro else [0, int(jogada)]
            # cria um novo registro de vitórias ordenado do mais vitorioso para o menso
            novos = sorted(registros + [(ganhos+1, casa)])
            # evita um bug que existe na versão 2.1.3 do brython
            novos = novos if len(novos) ==1 else novos[::-1]
            # Atualiza a história com esta nova lista de jogadas vitoriosas
            Tabuleiro.HISTORY[chave] = novos
        print(Tabuleiro.HISTORY)

    def clicou(self, event):
        cc = self.cena.mouse.pick().pos  # pega a posição do objeto clicado pelo mouse
        clicada = casa_clicada = (cc.x, cc.y, cc.z)  # cria uma tripla ordenada no espaço
        if casa_clicada in Casa.CASAS.keys():  # procura a tripla na coleção de casas
            casa_clicada = Casa.CASAS[casa_clicada]
            if casa_clicada in self.casas:
                self.remove(casa_clicada)
                if casa_clicada.clicou():
                    return # chama o clicou da casa escolhida
        #registra_jogada_na_historia(clicada)
        self.jogada(clicada)

TABULEIRO = None
def main():
    global TABULEIRO

    _gs = glow('main')
    cena = canvas()
    cena.width = 400
    cena.height = 400
    print(unquote(doc["history"].text))

    #TABULEIRO = [Casa(coluna, linha, camada)
    #             for coluna in TAM for linha in TAM for camada in TAM]
    TABULEIRO = Tabuleiro(cena)
    cena.bind("mousedown", TABULEIRO.clica)
    TABULEIRO.jogada()
    return TABULEIRO

if __name__ == "__main__":
    main()
